include/glow/Optimizer/Partition.h

```diff
@@ -0,0 +1,75 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef GLOW_OPTIMIZER_PARTITION_H
+#define GLOW_OPTIMIZER_PARTITION_H
+
+#include "glow/Graph/Graph.h"
+
+#include <llvm/ADT/DenseMap.h>
+
+namespace glow {
+
+/// Maps a set of functions to the set of functions it depends on.  The
+/// execution of a function cannot begin until all its dependencies have
+/// executed.
+class FunctionGraph {
```

Comment:
Maybe a better name would be FunctionDAG or FunctionDependencyDAG?

Reply:
How are you planning to AOT this class?

Reply:
How are you going to visualize this data structure?

Reply:
There are lots of options for AOT.  The trivial thing would be to handle it externally (like the serial JIT executor): dump N bundles, then simply invoke their exported functions serially.

To actually take advantage of the partitioning, I'd serialize the `FunctionGraph` into `BundleConfig` so we can reconstitute it as part of the network state.  From there I see at least two options:
1. Leave it up to the "glue" code to decide how to execute the graph.  To make this easier, we could provide a library of executors that's independent of the Glow core.
2. Compile-in the execution strategy by making the executor part of libjit.  That would take a bit more codegen to emit the glue, but makes the external interface easier.
3. (Maybe a bit of a research project) Drop the notion of an executor entirely and compile-in the dependences to the individual functions.

Reply:
Re: visualization, I've found it useful to dump the individual partitions to dot.  It would be extremely to cool to dump the whole thing to a single dot file that uses subgraphs to delineate the partitions, but the graphviz wizardry for that is pretty high level (if even possibly for complicate graphs).  Maybe I'll try it :-p.

Reply:
I wonder if turning FunctionGraph into a function with Call nodes would solve the AOT and the visualization problem. AOT would be done by emitting 'call' instructions, and you wull get visualization for free.

Reply:
Hm, interesting idea.  I kind of think the hardest part exists either way... for AOT you have to figure out how to compile a bunch of functions and then refer to them from some caller.  I don't see a huge gap between a function-dag (containing functions) and a function (containing call nodes).